### PR TITLE
Add script to check for absl reference from Firestore/core/api

### PR DESCRIPTION
**Pull request by wu-hui**
Also add exception to check_imports.swift to exclude cmake build directories.

### CODE REVIEW

File: scripts/check_firestore_core_api_absl.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  _logger.debug('Checking Firestore/core/src/api absl usage against %s' %
+                revision)
+  diff = diff_with_absl(revision, ['Firestore/core/src/api/*.h'])
+
+  # Check for changes adding new absl references only.
+  found = False
+  for line in diff.splitlines():
+    # Additions start with '+'
+    if line.startswith('+') and line.find('absl::') >= 0:
+      # Found a change introducing a reference to absl
```
Comment (var-const) [inline scripts/check_firestore_core_api_absl.py:75]: Nit: this comment repeats the message from the next line, I'd omit it.

File: scripts/check_firestore_core_api_absl.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  command.extend(patterns)
+  command_trace.log(command)
+  return six.ensure_text(subprocess.check_output(command))
+
+
+def main():
+  parser = argparse.ArgumentParser(description='Lint source files.')
+  parser.add_argument(
+      '--dry-run',
+      '-n',
```
Comment (var-const) [inline scripts/check_firestore_core_api_absl.py:49]: Question: what is the use for dry run? From what I can see, the main difference is that it always returns a successful code upon exit.

File: scripts/check_firestore_core_api_absl.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+      nargs='?',
+      help='A single revision that specifies a point in time '
+      'from which to look for changes. Defaults to '
+      'origin/master.')
+  args = command_trace.parse_args(parser)
+
+  dry_run = False
+  if args.dry_run:
+    dry_run = True
+    command_trace.enable_tracing()
```
Comment (var-const) [inline scripts/check_firestore_core_api_absl.py:63]: Question: what does this command do?

**Review by var-const — CHANGES_REQUESTED** [review]
Overall looking good, just some nitpicks.

File: scripts/check_firestore_core_api_absl.py (hunk 3 above)
Comment (wu-hui) [reply scripts/check_firestore_core_api_absl.py]: It just sets the log level to debug.

File: scripts/check_firestore_core_api_absl.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+  revision = 'origin/master' if (not args.rev) else args.rev
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply scripts/check_firestore_core_api_absl.py]: Done.

File: scripts/check_firestore_core_api_absl.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+  _logger.debug('Checking Firestore/core/src/api absl usage against %s' %
+                revision)
+  diff = diff_with_absl(revision, ['Firestore/core/src/api/*.h'])
+
+  # Check for changes adding new absl references only.
+  found = False
+  for line in diff.splitlines():
+    # Additions start with '+'
+    if line.startswith('+') and line.find('absl::') >= 0:
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply scripts/check_firestore_core_api_absl.py]: Done.

File: scripts/check_firestore_core_api_absl.py (hunk 1 above)
Comment (wu-hui) [reply scripts/check_firestore_core_api_absl.py]: Done.

File: scripts/check_firestore_core_api_absl.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+      # Found a change introducing a reference to absl
+      _logger.error(
+          'Found a change introducing reference to absl under'
+          'Firestore/core/api:'
+      )
+      _logger.error('  %s' % line)
+      _logger.error('')
+      found = True
+
+  if found and (not dry_run):
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply scripts/check_firestore_core_api_absl.py]: Done.

File: scripts/check_firestore_core_api_absl.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+  for line in diff.splitlines():
+    # Additions start with '+'
+    if line.startswith('+') and line.find('absl::') >= 0:
+      # Found a change introducing a reference to absl
+      _logger.error(
+          'Found a change introducing reference to absl under'
+          'Firestore/core/api:'
+      )
+      _logger.error('  %s' % line)
+      _logger.error('')
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply scripts/check_firestore_core_api_absl.py]: Done.

File: scripts/check_firestore_core_api_absl.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+      'origin/master.')
+  args = command_trace.parse_args(parser)
+
+  dry_run = False
+  if args.dry_run:
+    dry_run = True
+    command_trace.enable_tracing()
+
+  revision = 'origin/master' if (not args.rev) else args.rev
+  _logger.debug('Checking Firestore/core/src/api absl usage against %s' %
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply scripts/check_firestore_core_api_absl.py]: Done.

File: scripts/check_firestore_core_api_absl.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+                revision)
+  diff = diff_with_absl(revision, ['Firestore/core/src/api/*.h'])
+
+  # Check for changes adding new absl references only.
+  found = False
+  for line in diff.splitlines():
+    # Additions start with '+'
+    if line.startswith('+') and line.find('absl::') >= 0:
+      # Found a change introducing a reference to absl
+      _logger.error(
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply scripts/check_firestore_core_api_absl.py:75]: Done.

File: scripts/check_firestore_core_api_absl.py [hunk 10]
```diff
@@ -34,16 +34,16 @@ def diff_with_absl(revision, patterns):
...
   command.extend(patterns)
-  command_trace.log(command)
+  _logger.debug(command)
   return six.ensure_text(subprocess.check_output(command))
 
 
 def main():
-  parser = argparse.ArgumentParser(description='Lint source files.')
+  parser = argparse.ArgumentParser(
+      description='Check Absl usaging in firestore/core/api')
```
Comment (var-const) [inline scripts/check_firestore_core_api_absl.py:46]: Nits:
- `s/usaging/usage/`;
- please use the `patterns` variable instead of hardcoding `api/` here.

**Review by var-const — COMMENTED** [review]
Question: is this possible to make it so that if this script fails, the failure is visible but the PR can still be merged? It's ok to add Abseil to the private section of a class in `api/`, but this script would still fail because it doesn't distinguish between `public` and `private` (and trying to do that would be more trouble than it's worth, IMO).

**Comment by var-const** [comment]
Mostly LGTM, but need to resolve the question about whether this check blocks/should block merging.

File: scripts/check_firestore_core_api_absl.py (hunk 10 above)
Comment (wu-hui) [reply scripts/check_firestore_core_api_absl.py:46]: Done.

**Comment by wu-hui** [comment]
> Question: is this possible to make it so that if this script fails, the failure is visible but the PR can still be merged? It's ok to add Abseil to the private section of a class in `api/`, but this script would still fail because it doesn't distinguish between `public` and `private` (and trying to do that would be more trouble than it's worth, IMO).

We could change the call with `--dry-run` in `check.sh`. But we can wait until such a case did happen: when someone tried to add absl in private. I doubt this would happen before we remove this check.

**Review by var-const — APPROVED** [review]

**Comment by var-const** [comment]
@wu-hui Can you please do a test PR that adds Abseil to the `api` folder after you merge this? Just to make sure that the check fails and to take a look at the script output.

**Comment by var-const** [comment]
@wu-hui Andy, is this ready to merge? Please let me know if you need anything from me.

**Pull request merged**